Assets/LeapMotionModules/InteractionEngine/Scripts/ControllerImplementations/HeuristicGrabClassifier.cs

```diff
@@ -0,0 +1,142 @@
+using Leap.Unity.RuntimeGizmos;
+
+#if UNITY_EDITOR
+using UnityEditor;
+#endif
+
+namespace Leap.Unity.Interaction {
+  public class HeuristicGrabClassifier {
+    Collider[] collidingCandidates = new Collider[10];
+    public InteractionManager _manager;
+    private Hand _hand;
+
+    Dictionary<IInteractionBehaviour, GrabClassifier> leftGrabClassifiers = new Dictionary<IInteractionBehaviour, GrabClassifier>();
+    Dictionary<IInteractionBehaviour, GrabClassifier> rightGrabClassifiers = new Dictionary<IInteractionBehaviour, GrabClassifier>();
+
+    public HeuristicGrabClassifier(InteractionManager manager) {
+      _manager = manager;
+    }
+
+    //Grab Classifier Logic
+    public void UpdateClassifier(GrabClassifier classifier) {
+      classifier.warpTrans = Matrix4x4.TRS(classifier.warper.RigidbodyPosition, classifier.warper.RigidbodyRotation, classifier.transform.localScale);
+      
+      //For each probe (fingertip)
+      for (int j = 0; j < classifier.probes.Length; j++) {
+        //Calculate how extended the finger is
+        float tempCurl = Vector3.Dot(_hand.Fingers[j].Direction.ToVector3(), (j != 0) ? _hand.Direction.ToVector3() : (_hand.IsLeft ? 1f : -1f) * _hand.Basis.xBasis.ToVector3());
+
+        //Determine if this probe is intersecting an object
+        Array.Clear(collidingCandidates, 0, 10);
+        Physics.OverlapSphereNonAlloc(_hand.Fingers[j].TipPosition.ToVector3(), j == 0 ? 0.015f : 0.01f, collidingCandidates);
+        bool collidingWithObject = false;
+        foreach(Collider col in collidingCandidates) {
+          if (col!=null && col.attachedRigidbody != null) {
+            collidingWithObject = (col.attachedRigidbody == classifier.body) ? true : collidingWithObject;
+          }
+        }
+
+        //Nullify above findings if fingers are extended
+        bool tempIsInside = collidingWithObject && (tempCurl < 0.65f) && (tempCurl > -0.1f);
+
+        //Probes go inside when they intersect, probes come out when they uncurl
+        if (!classifier.probes[j].isInside) {
+          classifier.probes[j].isInside = tempIsInside;
+          classifier.probes[j].curl = tempCurl + (j==0?0.15f:0f);
+        } else {
+          if (tempCurl > classifier.probes[j].curl) {
+            classifier.probes[j].isInside = tempIsInside;
+          }
+        }
+      }
+
+      //If thumb and one other finger is "inside" the object, it's a grab!
+      classifier.isGrabbing = (classifier.probes[0].isInside && (classifier.probes[1].isInside ||
+                                                                 classifier.probes[2].isInside ||
+                                                                 classifier.probes[3].isInside ||
+                                                                 classifier.probes[4].isInside));
+      //If grabbing within 10 frames of releasing, discard grab.
+      //Suppresses spurious regrabs and makes throws work better.
+      if (classifier.coolDown <= 10) {
```

Comment:
I would definitely want this to be time based instead of frame based.  With our work on mobile we are going to be having different frame-rates across different applications, both for rendering and for physics.